pywhy_stats/kci.py

```diff
@@ -0,0 +1,393 @@
+
+
+def _ind(Kx, Ky, approx, null_sample_size, threshold):
+    # test statistic is the normal bivariate independence test
+    test_stat = _compute_V_statistic(Kx, Ky)
+
+    if approx:
+        # approximate the pvalue using the Gamma distribution
+        k_appr, theta_appr = _approx_gamma_params_ind(Kx, Ky)
+        pvalue = 1 - stats.gamma.cdf(test_stat, k_appr, 0, theta_appr)
+    else:
+        null_samples = _compute_null_ind(Kx, Ky, null_sample_size, 1000, threshold)
+        pvalue = np.sum(null_samples > test_stat) / float(null_sample_size)
+
+    return test_stat, pvalue
+
+
+def _cond(Kx, Ky, Kz, approx, null_sample_size, threshold):
+    # compute the centralizing matrix for the kernels according to
+    # conditioning set Z
+    epsilon = 1e-6
+    n = Kx.shape[0]
+    Rz = epsilon * np.linalg.pinv(Kz + epsilon * np.eye(n))
+
+    # compute the centralized kernel matrices
+    KxzR = Rz.dot(Kx).dot(Rz)
+    KyzR = Rz.dot(Ky).dot(Rz)
+
+    # compute the conditional independence test statistic
+    test_stat = _compute_V_statistic(KxzR, KyzR)
+
+    # compute the product of the eigenvectors
+    uu_prod = _compute_prod_eigvecs(KxzR, KyzR, threshold=threshold)
+
+    if approx:
+        # approximate the pvalue using the Gamma distribution
+        k_appr, theta_appr = _approx_gamma_params_ci(uu_prod)
+        pvalue = 1 - stats.gamma.cdf(test_stat, k_appr, 0, theta_appr)
+    else:
+        null_samples = _compute_null_ci(uu_prod, null_sample_size, threshold)
+        pvalue = np.sum(null_samples > test_stat) / float(null_sample_size)
+
+    return test_stat, pvalue
+
+
+def _compute_V_statistic(KxR, KyR):
+    # n = KxR.shape[0]
+    # compute the sum of the two kernsl
+    Vstat = np.sum(KxR * KyR)
+    return float(Vstat)
+
+
+def _approx_gamma_params_ind(Kx, Ky):
+    T = Kx.shape[0]
+    mean_appr = np.trace(Kx) * np.trace(Ky) / T
+    var_appr = 2 * np.trace(Kx.dot(Kx)) * np.trace(Ky.dot(Ky)) / T / T
+    k_appr = mean_appr**2 / var_appr
+    theta_appr = var_appr / mean_appr
+    return k_appr, theta_appr
+
+
+def _compute_null_ind(Kx, Ky, n_samples, max_num_eigs, threshold):
+    n = Kx.shape[0]
+
+    # get the eigenvalues in ascending order, smallest to largest
+    eigvals_x = np.linalg.eigvalsh(Kx)
+    eigvals_y = np.linalg.eigvalsh(Ky)
+
+    # optionally only keep the largest "N" eigenvalues
+    eigvals_x = eigvals_x[-max_num_eigs:]
+    eigvals_y = eigvals_y[-max_num_eigs:]
+    num_eigs = len(eigvals_x)
+
+    # compute the entry-wise product of the eigenvalues and store it as a vector
+    eigvals_prod = np.dot(eigvals_x.reshape(num_eigs, 1), eigvals_y.reshape(1, num_eigs)).reshape(
+        (-1, 1)
+    )
+
+    # only keep eigenvalues above a certain threshold
+    eigvals_prod = eigvals_prod[eigvals_prod > eigvals_prod.max() * threshold]
+
+    # generate chi-square distributed values z_{ij} with degree of freedom 1
+    f_rand = np.random.chisquare(df=1, size=(len(eigvals_prod), n_samples))
```

Comment:
For all the private functions, we end up using `np.random`. This means we don't have any way of promising users that we have control over sequential runs on data. For all the private functions, we can pass in `rng=None` and in the ind/condind functions, we can pass in `seed=None`. https://numpy.org/doc/stable/reference/random/generator.html

At the top of the functions, we can just the pattern:

```
def ind(...., seed=None):
      # if seed is None, then every run is different. 
      rng = np.random.default_rng(seed)

      # calling private functions
      _compute_null_ind(..., rng=rng)
```

Reply:
No need for that, it will be deterministic in sequential runs by simply setting the seed outside:
```
np.random.seed(0)
ind(X,Y)
np.random.seed(0)
ind(X,Y)
```
This will produce the exact same result (i.e., it is deterministic). I find this cleaner than passing a random seed, seeing that one typically would set a random seed outside to make an algorithm deterministic.

Reply:
This is in general bad practice because it can lead to a bunch of errors that we cannot control. I realize this might be a shift in coding patterns, but I've ran into this issue a bunch over the years and its jaded me.

In this pattern, we'll have to explicitly inform the user to set the random seed in such a way and take the responsibility of educating them on how to get reproducibility. This could also result in a bunch of issues in terms of reproducibility for the novice user (and several errors when we transition to parallelized calls), which I think we would like to do when we eat hyppo package (a lot of code is Cythonized, which will require passing in an instance of random number generator since `np.random` is not thread safe):

- https://stackoverflow.com/questions/31057197/should-i-use-random-seed-or-numpy-random-seed-to-control-random-number-gener
- https://builtin.com/data-science/numpy-random-seed

In your example:

E.g. the second run is not reproducible if there is some stochastic parameter. E.g. depending on what LOC is run that may involve calls to the random states, this may change the result even tho the user wants the "seed to be the same"? 
```
np.random.seed(0)
ind(X,Y)
np.random.seed(0)
ind(X,Y, option=stochastic)
```

And if a user runs:
```
np.random.seed(0)
ind(X,Y)
...
ind(X, Y)
```

they're probably going to complain about the runs not being reproducible.

Reply:
Regarding parallelization, that is not an issue, since you can deterministically create random seeds for the parallel processes, see e.g. https://github.com/py-why/dowhy/blob/main/dowhy/gcm/shapley.py#L220 where we do something similar. In that way, even parallel processes are deterministic.

In the KCI case here, the second run is the same as the first one. Generally, there can indeed be the case that not all functionalities rely on the same random generator, for this we use a certain function in dowhy for instance to make sure that this happens (see https://github.com/py-why/dowhy/blob/961ffc63732982bec12503926c6077c8daec7dcc/dowhy/gcm/util/general.py#L44)

That being said, if you think this really causes any confusion, I am also ok with offering this random seed parameter, since one can still go the route I mentioned by setting it externally and leave the parameter blank. At least it makes it clearer for people who are used to these parameters.

Reply:
I think most of our users will come from researchers who use sklearn/numpy, so in favor of offering the seed parameter, so they know we can take care of the randomness for them.